File: python/2020/day_02/main.py
import re


def get_input(path_name="./input.txt"):
    with open(path_name) as f:
        return list(x.strip() for x in f.readlines())


def part_1():
    regex = '(-| |: )'
    parsed_input = [(int(re.split(regex, x)[0]), int(re.split(regex, x)[
                     2]), re.split(regex, x)[4], re.split(regex, x)[-1]) for x in get_input()]

    count = 0
    for minimum, maximum, letter, sequence in parsed_input:
        if sequence.count(letter) >= minimum and sequence.count(letter) <= maximum:
            count += 1
    return count


def part_2():
    regex = '(-| |: )'
    parsed_input = [(int(re.split(regex, x)[0]), int(re.split(regex, x)[
                     2]), re.split(regex, x)[4], re.split(regex, x)[-1]) for x in get_input()]

    count = 0
    for first_loc, second_loc, letter, sequence in parsed_input:
        if sequence[first_loc - 1] == letter and sequence[second_loc - 1] == letter:
            pass
        elif sequence[first_loc - 1] == letter:
            count += 1
        elif sequence[second_loc - 1] == letter:
            count += 1
        else:
            pass

    return count


if __name__ == "__main__":
    print(part_1())  # 638
    print(part_2())  # 699

File: rust/2020/day_02/src/main.rs
use regex::Regex;
use std::fs::read_to_string;

#[derive(Debug, Clone)]
struct Password {
    lower: usize,
    upper: usize,
    character: char,
    password: String,
}

impl Password {
    fn verify(&self) -> bool {
        let mut count = 0;
        for c in self.password.chars() {
            if c == self.character {
                count += 1;
            }
        }
        return count >= self.lower && count <= self.upper;
    }

    fn verify_two(&self) -> bool {
        let lower = self.password.chars().nth(self.lower - 1).unwrap() == self.character;
        let upper = self.password.chars().nth(self.upper - 1).unwrap() == self.character;
        return upper ^ lower;
    }
}

fn main() {
    part_1();
    part_2();
}

fn parse() -> Vec<Password> {
    let input = read_to_string("../input.txt").unwrap();
    let mut passwords: Vec<Password> = vec![];

    let re: Regex = Regex::new(r"(\d+)-(\d+) (\w): (\w+)").unwrap();
    for line in input.lines() {
        for cap in re.captures_iter(line) {
            passwords.push(Password {
                lower: cap[1].parse::<usize>().unwrap(),
                upper: cap[2].parse::<usize>().unwrap(),
                character: cap[3].parse::<char>().unwrap(),
                password: cap[4].to_string(),
            });
        }
    }
    passwords
}

fn part_1() {
    let mut count = 0;
    let passwords = parse();
    for p in passwords {
        if p.part_1() {
            count += 1;
        }
    }
    println!("part 1's count is: {}", count);
}

fn part_2() {
    let mut count = 0;
    let passwords = parse();
    for p in passwords {
        if p.part_two() {
            count += 1;
        }
    }
    println!("part 2's count is: {}", count);
}
